### PR TITLE
chuong 5

### PR DESCRIPTION
**Pull request by huuhau95**

### CODE REVIEW

File: app/assets/javascripts/static_pages.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (mdmahmudurrahman) [inline app/assets/javascripts/static_pages.coffee]: Please remove this unused file.

File: app/assets/javascripts/users.coffee [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (mdmahmudurrahman) [inline app/assets/javascripts/users.coffee]: Please remove this unused file.

File: app/assets/stylesheets/custom.scss [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  font-weight: normal;
+  color: $gray-light;
+}
+
+p {
+  font-size: 1.1em;
+  line-height: 1.7em;
+}
+
+
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Please remove this blank line.

File: app/assets/stylesheets/static_pages.scss [hunk 4]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the StaticPages controller here.
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/static_pages.scss]: Please remove this unused file.

File: app/assets/stylesheets/users.scss [hunk 5]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Users controller here.
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/users.scss]: Please remove this unused file.

File: app/helpers/application_helper.rb [hunk 6]
```diff
@@ -1,2 +1,12 @@
 module ApplicationHelper
-end
+
+  # Returns the full title on a per-page basis.
```
Comment (mdmahmudurrahman) [inline app/helpers/application_helper.rb]: Please remove this comment.

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -1,2 +1,12 @@
...
+  def full_title(page_title = '')
+    base_title = "Ruby on Rails Tutorial Sample App"
```
Comment (mdmahmudurrahman) [inline app/helpers/application_helper.rb]: Please use [Rails Internationalization (I18n) API](http://guides.rubyonrails.org/i18n.html)

File: app/helpers/static_pages_helper.rb [hunk 8]
```diff
@@ -0,0 +1,2 @@
+module StaticPagesHelper
```
Comment (mdmahmudurrahman) [inline app/helpers/static_pages_helper.rb]: Please remove this unused file.

File: app/helpers/users_helper.rb [hunk 9]
```diff
@@ -0,0 +1,2 @@
+module UsersHelper
+end
```
Comment (mdmahmudurrahman) [inline app/helpers/users_helper.rb]: Please remove this unused file.

File: app/views/layouts/_footer.html.erb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+<footer class="footer">
+  <small>
+    The <%= link_to "Ruby on Rails Tutorial", "http://www.railstutorial.org/" %>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_footer.html.erb]: Please use [Rails Internationalization (I18n) API](http://guides.rubyonrails.org/i18n.html
)

File: app/views/layouts/_footer.html.erb [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+    by <%= link_to "Michael Hartl", "http://www.michaelhartl.com/" %>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_footer.html.erb]: Please use [Rails Internationalization (I18n) API](http://guides.rubyonrails.org/i18n.html
)

File: app/views/layouts/_footer.html.erb [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+  </small>
+  <nav>
+    <ul>
+      <li><%= link_to "About", about_path %></li>
+      <li><%= link_to "Contact", contact_path %></li>
+      <li><%= link_to "News", "http://news.railstutorial.org/" %></li>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_footer.html.erb]: I18n in all of link in this file.

File: app/views/layouts/_header.html.erb [hunk 13]
```diff
@@ -0,0 +1,12 @@
+<header class="navbar navbar-fixed-top navbar-inverse">
+  <div class="container">
+    <%= link_to "sample app", root_path, id: "logo" %>
+    <nav>
+      <ul class="nav navbar-nav navbar-right">
+        <li><%= link_to "Home", root_path %></li>
+        <li><%= link_to "Help", help_path %></li>
+        <li><%= link_to "Log in", '#' %></li>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_header.html.erb]: I18n in all of link text in this file.

File: app/views/layouts/application.html.erb [hunk 14]
```diff
@@ -1,14 +1,22 @@
 <!DOCTYPE html>
 <html>
-  <head>
-    <title>Haugreek</title>
+
```
Comment (mdmahmudurrahman) [inline app/views/layouts/application.html.erb]: Remove the blank line

File: app/views/layouts/application.html.erb [hunk 15]
```diff
@@ -1,14 +1,22 @@
...
+</head>
 
-    <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload' %>
-    <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
-  </head>
+<body>
+    <%= render "layouts/header" %>
+    <div class="container">
+        <%= yield %>
+        <%= render "layouts/footer" %>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/application.html.erb]: 2 space indentation

File: app/views/layouts/application.html.erb [hunk 16]
```diff
@@ -1,14 +1,22 @@
...
+    <%= javascript_include_tag "application", "data-turbolinks-track": "reload" %>
+    <%= render "layouts/shim" %>
+</head>
 
-    <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload' %>
-    <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
-  </head>
+<body>
+    <%= render "layouts/header" %>
+    <div class="container">
```
Comment (mdmahmudurrahman) [inline app/views/layouts/application.html.erb]: 2 space indentation

File: app/views/layouts/application.html.erb [hunk 17]
```diff
@@ -1,14 +1,22 @@
 <!DOCTYPE html>
 <html>
-  <head>
-    <title>Haugreek</title>
+
+<head>
+    <title>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/application.html.erb]: 2 space indentation

File: app/views/layouts/application.html.erb [hunk 18]
```diff
@@ -1,14 +1,22 @@
...
+        <%= full_title(yield(:title)) %>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/application.html.erb]: 2 space indentation

File: app/views/static_pages/about.html.erb [hunk 19]
```diff
@@ -0,0 +1,10 @@
+<% provide(:title, "About") %>
+<h1>About</h1>
```
Comment (mdmahmudurrahman) [inline app/views/static_pages/about.html.erb]: I18n

File: app/views/static_pages/about.html.erb [hunk 20]
```diff
@@ -0,0 +1,10 @@
...
+<p>
+  The <%= link_to "http://www.railstutorial.org/" do %>
+  <%= content_tag :em, "Ruby on Rails" %>
+  <% end %> is a
+  <%= link_to "book", "http://www.railstutorial.org/book" %> and
+  <%= link_to "screencast series", "http://screencasts.railstutorial.org/" %> to teach web development with
+  <%= link_to "Ruby on Rails", "http://rubyonrails.org/" %> This is the sample application for the tutorial.
```
Comment (mdmahmudurrahman) [inline app/views/static_pages/about.html.erb]: Please use I18n in all of the text for multi-language support. You can visit the following link to do this. 
[Rails Internationalization (I18n) API](http://guides.rubyonrails.org/i18n.html
)

File: app/views/static_pages/contact.html.erb [hunk 21]
```diff
@@ -0,0 +1,6 @@
+<% provide(:title, 'Contact') %>
```
Comment (mdmahmudurrahman) [inline app/views/static_pages/contact.html.erb]: Please use double quote.

File: app/views/static_pages/contact.html.erb [hunk 22]
```diff
@@ -0,0 +1,6 @@
...
+<h1>Contact</h1>
+<p>
+  Contact the Ruby on Rails Tutorial about the sample app at the
+  <%= link_to "contact page.", "http://www.railstutorial.org/contact" %>
```
Comment (mdmahmudurrahman) [inline app/views/static_pages/contact.html.erb]: Please use I18n for multi-language support

File: app/views/static_pages/help.html.erb [hunk 23]
```diff
@@ -0,0 +1,11 @@
+<% provide(:title, "Help") %>
+<h1>Help</h1>
+<p>
+  Get help on the Ruby on Rails Tutorial at the
+  <%= link_to "Rails Tutorial help section", "http://www.railstutorial.org/help" %>.
+  To get help on this sample app, see the
+  <%= link_to "", "http://www.railstutorial.org/book" %>
+  <%= link_to "http://www.railstutorial.org/book" do %>
+  <%= content_tag :em, "Ruby on Rails Tutorial." %>
+  <% end %>
```
Comment (mdmahmudurrahman) [inline app/views/static_pages/help.html.erb]: Same comment for the I18n in this file.

File: app/views/static_pages/home.html.erb [hunk 24]
```diff
@@ -0,0 +1,10 @@
+<div class="center jumbotron">
+  <h1>Welcome to the Sample App</h1>
+  <h2>
+    This is the home page for the
+    <a href="http://www.railstutorial.org/">Ruby on Rails Tutorial</a>
+    sample application.
+  </h2>
+  <%= link_to "Sign up now!", signup_path, class: "btn btn-lg btn-primary" %>
+</div>
+<%= link_to image_tag("rails.png", alt: "Rails logo"), 'http://rubyonrails.org/' %>
```
Comment (mdmahmudurrahman) [inline app/views/static_pages/home.html.erb]: Same comment for the I18n in this file and use double quote.

File: app/views/users/new.html.erb [hunk 25]
```diff
@@ -0,0 +1,3 @@
+<% provide(:title, 'Sign up') %>
```
Comment (mdmahmudurrahman) [inline app/views/users/new.html.erb]: Use double quote.

File: app/views/users/new.html.erb [hunk 26]
```diff
@@ -0,0 +1,3 @@
...
+<h1>Sign up</h1>
+<p>This will be a signup page for new users.</p>
```
Comment (mdmahmudurrahman) [inline app/views/users/new.html.erb]: I18n

File: app/helpers/application_helper.rb [hunk 27]
```diff
@@ -1,2 +1,12 @@
...
+    if page_title.empty?
+      base_title
+    else
+      page_title + " | " + base_title
+    end
+  end
+end
```
Comment (mdmahmudurrahman) [inline app/helpers/application_helper.rb]: Unavailable end of line.

**Comment by mdmahmudurrahman** [comment]
commented

File: app/assets/javascripts/users.coffee (hunk 2 above)
Comment (huuhau95) [reply app/assets/javascripts/users.coffee]: done

File: app/assets/stylesheets/custom.scss (hunk 3 above)
Comment (huuhau95) [reply app/assets/stylesheets/custom.scss]: done

File: app/assets/stylesheets/static_pages.scss (hunk 4 above)
Comment (huuhau95) [reply app/assets/stylesheets/static_pages.scss]: done

File: app/assets/stylesheets/users.scss (hunk 5 above)
Comment (huuhau95) [reply app/assets/stylesheets/users.scss]: done

File: app/helpers/application_helper.rb (hunk 6 above)
Comment (huuhau95) [reply app/helpers/application_helper.rb]: done

File: app/helpers/static_pages_helper.rb (hunk 8 above)
Comment (huuhau95) [reply app/helpers/static_pages_helper.rb]: done

File: app/helpers/users_helper.rb (hunk 9 above)
Comment (huuhau95) [reply app/helpers/users_helper.rb]: done

File: app/views/layouts/_footer.html.erb (hunk 10 above)
Comment (huuhau95) [reply app/views/layouts/_footer.html.erb]: done

File: app/views/layouts/_footer.html.erb (hunk 11 above)
Comment (huuhau95) [reply app/views/layouts/_footer.html.erb]: done

File: app/views/layouts/_header.html.erb (hunk 13 above)
Comment (huuhau95) [reply app/views/layouts/_header.html.erb]: done

File: app/views/layouts/application.html.erb (hunk 14 above)
Comment (huuhau95) [reply app/views/layouts/application.html.erb]: done

File: app/views/layouts/application.html.erb (hunk 15 above)
Comment (huuhau95) [reply app/views/layouts/application.html.erb]: done

File: app/views/layouts/application.html.erb (hunk 16 above)
Comment (huuhau95) [reply app/views/layouts/application.html.erb]: done

File: app/views/layouts/application.html.erb (hunk 17 above)
Comment (huuhau95) [reply app/views/layouts/application.html.erb]: done

File: app/views/users/new.html.erb (hunk 26 above)
Comment (huuhau95) [reply app/views/users/new.html.erb]: done

File: app/views/users/new.html.erb (hunk 25 above)
Comment (huuhau95) [reply app/views/users/new.html.erb]: done

File: app/views/static_pages/home.html.erb (hunk 24 above)
Comment (huuhau95) [reply app/views/static_pages/home.html.erb]: done

File: app/views/static_pages/help.html.erb (hunk 23 above)
Comment (huuhau95) [reply app/views/static_pages/help.html.erb]: done

File: app/views/static_pages/contact.html.erb (hunk 22 above)
Comment (huuhau95) [reply app/views/static_pages/contact.html.erb]: done

File: app/views/static_pages/contact.html.erb (hunk 21 above)
Comment (huuhau95) [reply app/views/static_pages/contact.html.erb]: done

File: app/views/static_pages/about.html.erb (hunk 20 above)
Comment (huuhau95) [reply app/views/static_pages/about.html.erb]: done

File: app/views/static_pages/about.html.erb (hunk 19 above)
Comment (huuhau95) [reply app/views/static_pages/about.html.erb]: done

File: app/views/layouts/application.html.erb (hunk 18 above)
Comment (huuhau95) [reply app/views/layouts/application.html.erb]: done

File: app/helpers/application_helper.rb (hunk 7 above)
Comment (huuhau95) [reply app/helpers/application_helper.rb]: done

File: app/views/layouts/_footer.html.erb (hunk 12 above)
Comment (huuhau95) [reply app/views/layouts/_footer.html.erb]: done

**Comment by huuhau95** [comment]
Ready

**Comment by huuhau95** [comment]
ready

**Comment by Tranducthang28011996** [comment]
commented

File: app/controllers/static_pages_controller.rb [hunk 28]
```diff
@@ -0,0 +1,8 @@
+class StaticPagesController < ApplicationController
+  def home
+  end
+  def help
+  end
+  def about
+  end
+end
```
[the remark this answers is not in the document]
Comment (huuhau95) [reply app/controllers/static_pages_controller.rb]: ok b

File: Gemfile [hunk 29]
```diff
@@ -1,54 +1,57 @@
...
-  gem 'spring'
-  gem 'spring-watcher-listen', '~> 2.0.0'
+  gem "spring"
+  gem "spring-watcher-listen", "~> 2.0.0"
 end
 
 # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
-gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
+gem "tzinfo-data", platforms: [:mingw, :mswin, :x64_mingw, :jruby]
+
```
[the remark this answers is not in the document]
Comment (huuhau95) [reply Gemfile]: done

File: Guardfile [hunk 30]
```diff
@@ -0,0 +1,55 @@
...
+    resource_tests(matches[1])
+  end
+  watch(%r{^app/views/([^/]*?)/.*\.html\.erb$}) do |matches|
+    ["test/controllers/#{matches[1]}_controller_test.rb"] +
+    integration_tests(matches[1])
+  end
+  watch(%r{^app/helpers/(.*?)_helper\.rb$}) do |matches|
+    integration_tests(matches[1])
+  end
+  watch('app/views/layouts/application.html.erb') do
```
[the remark this answers is not in the document]
Comment (huuhau95) [reply Guardfile]: done

File: Guardfile [hunk 31]
```diff
@@ -0,0 +1,55 @@
...
+    'test/integration/site_layout_test.rb'
+  end
+  watch('app/helpers/sessions_helper.rb') do
+    integration_tests << 'test/helpers/sessions_helper_test.rb'
+  end
+  watch('app/controllers/sessions_controller.rb') do
```
[the remark this answers is not in the document]
Comment (huuhau95) [reply Guardfile]: done

File: Guardfile [hunk 32]
```diff
@@ -0,0 +1,55 @@
...
+
+# Returns the controller tests corresponding to the given resource.
+def controller_test(resource)
+  "test/controllers/#{resource}_controller_test.rb"
+end
+
+# Returns all tests for the given resource.
+def resource_tests(resource)
+  integration_tests(resource) << controller_test(resource)
+end
```
[the remark this answers is not in the document]
Comment (huuhau95) [reply Guardfile]: done

**Comment by huuhau95** [comment]
done

**Pull request merged**